### ISSUE TITLE
Add public copy constructors for HTTP header collections

### ISSUE THREAD

**Issue by Zastai** [dotnet/runtime#95912]
### Is your feature request related to a problem? Please describe.

I wanted to store the (content and response) headers for an HTTP response message (passed to me as a parameter so its lifetime is not under my control).

It's easy enough to keep a reference to them, but that has two issues:
1. Those collections are mutable; if my caller modifies them after calling me, my "saved" headers are also changed
2. The response message is disposable, and there is no guarantee (at least not in the API docs as far as I could see) that these header collections are safe to access once the response message is disposed.

Trying to keep a copy runs into issues:
- No public constructor for `HttpContentHeaders` or `HttpResponseHeaders`
- Both types are sealed, so I can't add a helper subclass either

### Describe the solution you'd like

I would like to be able to get a copy of those header collections.

Whether that is via a method on the existing instance (`response.Headers.CreateACopy()` or possibly making `HttpHeaders` implement `ICloneable`), a factory method (`HttpResponseHeaders.CopyOf(response.Headers)`) or a copy constructor (`new HttpContentHeaders(response.Content.Headers)` doesn't matter much.

Being able to choose between a mutable and immutable result would be even better (in my use case there's no good reason to allow mutating the saved copy).
A `bool mutable = true` argument to the copy method would likely work fine; adding `IHttpXxxHeaders` and `IReadOnlyHttpXxxHeaders` to make them act more like standard collections (while keeping their extra properties) would likely be excessive (but nice and consistent).

Note: if `MemberwiseClone` is intended to be safe to use for these header collections, I suppose that would be enough - but then that should be clearly mentioned in the class docs.

**Comment by CarnaViire** [comment]
How exactly do you intend to use the copy @Zastai? Is there a reason it has to be an instance of e.g. `HttpResponseHeaders`?

`HttpResponseHeaders` class (and all classes derived from `HttpHeaders`) implements `IEnumerable<KeyValuePair<string,IEnumerable<string>>>`, so you can easily copy it into e.g. a dictionary with a simple `ToDictionary()`/`ToImmutableDictionary()` call.

**Comment by Zastai** [comment]
Yes, I know I can preserve them as a dictionary, but that would drop the convenience properties that the specific classes provide. Keeping the type would also make my saved copies easier to consume by existing methods.

**Comment by MihaZupan** [comment]
FWIW you can create new header collections by creating new `HttpRequestMessage`/`HttpResponseMessage `objects.
Something like:

```c#
internal static class Helpers
{
    public static HttpRequestHeaders Copy(this HttpRequestHeaders headers)
    {
        var copy = new HttpRequestMessage().Headers;
        CopyHeaders(headers, copy);
        return copy;
    }

    public static HttpResponseHeaders Copy(this HttpResponseHeaders headers)
    {
        var copy = new HttpResponseMessage().Headers;
        CopyHeaders(headers, copy);
        return copy;
    }

    public static HttpContentHeaders Copy(this HttpContentHeaders headers)
    {
        var copy = new ByteArrayContent(Array.Empty<byte>()).Headers;
        CopyHeaders(headers, copy);
        return copy;
    }

    private static void CopyHeaders(HttpHeaders source, HttpHeaders destination)
    {
        foreach (var (key, value) in source.NonValidated)
        {
            destination.TryAddWithoutValidation(key, value);
        }
    }
}
```

**Comment by Zastai** [comment]
That's suboptimal, but it's one way of doing it.

It does run afoul of one thing: there is currently no defined contract that says that such a collection created as part of, say, `HttpResponseMessage` is not tied to that message, and remains valid when the message is disposed. (Is that why the sample code does not use `using` for the messages?)

**Comment by MihaZupan** [comment]
AFAIK The only connection that currently exists is between an `HttpContent` and its `Headers` collection. Specifically, if the `Content-Length` header has not been set, we will try to compute the value from the `HttpContent`.

Consider:
```c#
var content = new StreamContent(new MemoryStream(new byte[] { 1, 2, 3 }));
var headers = content.Headers;
_ = headers.ContentLength; // If you remove this, querying ContentLength later will throw
content.Dispose();
Console.WriteLine(headers.ContentLength);
```

So while in all other cases the collection is usable on its own, you don't have the guarantee that that will always be the case.

**Comment by MihaZupan** [comment]
Triage: We are open to considering exposing public constructors, but would like to see how much demand there is for such an API (please upvote the original post if this applies to you). Moving to Future.
